### PR TITLE
feat: improve engine spec discoverability

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

This PR changes the mechanism we use to discover DB engine specs, switching to [Python entry points](https://packaging.python.org/specifications/entry-points/). Using entry points is cleaner, and allows users to distribute new engine specs as packages separate from Superset.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

List of engines before:

```
{'awsathena': <class 'superset.db_engine_specs.athena.AthenaEngineSpec'>,
 'base': <class 'superset.db_engine_specs.base.BaseEngineSpec'>,
 'bigquery': <class 'superset.db_engine_specs.bigquery.BigQueryEngineSpec'>,
 'clickhouse': <class 'superset.db_engine_specs.clickhouse.ClickHouseEngineSpec'>,
 'cockroachdb': <class 'superset.db_engine_specs.cockroachdb.CockroachDbEngineSpec'>,
 'crate': <class 'superset.db_engine_specs.crate.CrateEngineSpec'>,
 'databricks': <class 'superset.db_engine_specs.databricks.DatabricksHiveEngineSpec'>,
 'dremio': <class 'superset.db_engine_specs.dremio.DremioBaseEngineSpec'>,
 'drill': <class 'superset.db_engine_specs.drill.DrillEngineSpec'>,
 'druid': <class 'superset.db_engine_specs.druid.DruidEngineSpec'>,
 'elasticsearch': <class 'superset.db_engine_specs.elasticsearch.ElasticSearchEngineSpec'>,
 'exa': <class 'superset.db_engine_specs.exasol.ExasolEngineSpec'>,
 'firebird': <class 'superset.db_engine_specs.firebird.FirebirdEngineSpec'>,
 'gsheets': <class 'superset.db_engine_specs.gsheets.GSheetsEngineSpec'>,
 'hana': <class 'superset.db_engine_specs.hana.HanaEngineSpec'>,
 'hive': <class 'superset.db_engine_specs.hive.HiveEngineSpec'>,
 'ibm_db_sa': <class 'superset.db_engine_specs.db2.Db2EngineSpec'>,
 'impala': <class 'superset.db_engine_specs.impala.ImpalaEngineSpec'>,
 'kylin': <class 'superset.db_engine_specs.kylin.KylinEngineSpec'>,
 'mssql': <class 'superset.db_engine_specs.mssql.MssqlEngineSpec'>,
 'mysql': <class 'superset.db_engine_specs.mysql.MySQLEngineSpec'>,
 'odelasticsearch': <class 'superset.db_engine_specs.elasticsearch.OpenDistroEngineSpec'>,
 'oracle': <class 'superset.db_engine_specs.oracle.OracleEngineSpec'>,
 'pinot': <class 'superset.db_engine_specs.pinot.PinotEngineSpec'>,
 'postgres': <class 'superset.db_engine_specs.postgres.PostgresEngineSpec'>,
 'postgresql': <class 'superset.db_engine_specs.postgres.PostgresEngineSpec'>,
 'presto': <class 'superset.db_engine_specs.presto.PrestoEngineSpec'>,
 'redshift': <class 'superset.db_engine_specs.redshift.RedshiftEngineSpec'>,
 'snowflake': <class 'superset.db_engine_specs.snowflake.SnowflakeEngineSpec'>,
 'solr': <class 'superset.db_engine_specs.solr.SolrEngineSpec'>,
 'sqlite': <class 'superset.db_engine_specs.sqlite.SqliteEngineSpec'>,
 'teradata': <class 'superset.db_engine_specs.teradata.TeradataEngineSpec'>,
 'trino': <class 'superset.db_engine_specs.trino.TrinoEngineSpec'>,
 'vertica': <class 'superset.db_engine_specs.vertica.VerticaEngineSpec'>}
```

And after:

```
{'awsathena': <class 'superset.db_engine_specs.athena.AthenaEngineSpec'>,
 'bigquery': <class 'superset.db_engine_specs.bigquery.BigQueryEngineSpec'>,
 'clickhouse': <class 'superset.db_engine_specs.clickhouse.ClickHouseEngineSpec'>,
 'cockroachdb': <class 'superset.db_engine_specs.cockroachdb.CockroachDbEngineSpec'>,
 'crate': <class 'superset.db_engine_specs.crate.CrateEngineSpec'>,
 'databricks': <class 'superset.db_engine_specs.databricks.DatabricksHiveEngineSpec'>,
 'dremio': <class 'superset.db_engine_specs.dremio.DremioBaseEngineSpec'>,
 'drill': <class 'superset.db_engine_specs.drill.DrillEngineSpec'>,
 'druid': <class 'superset.db_engine_specs.druid.DruidEngineSpec'>,
 'elasticsearch': <class 'superset.db_engine_specs.elasticsearch.ElasticSearchEngineSpec'>,
 'exa': <class 'superset.db_engine_specs.exasol.ExasolEngineSpec'>,
 'firebird': <class 'superset.db_engine_specs.firebird.FirebirdEngineSpec'>,
 'gsheets': <class 'superset.db_engine_specs.gsheets.GSheetsEngineSpec'>,
 'hana': <class 'superset.db_engine_specs.hana.HanaEngineSpec'>,
 'hive': <class 'superset.db_engine_specs.hive.HiveEngineSpec'>,
 'ibm_db_sa': <class 'superset.db_engine_specs.db2.Db2EngineSpec'>,
 'impala': <class 'superset.db_engine_specs.impala.ImpalaEngineSpec'>,
 'kylin': <class 'superset.db_engine_specs.kylin.KylinEngineSpec'>,
 'mssql': <class 'superset.db_engine_specs.mssql.MssqlEngineSpec'>,
 'mysql': <class 'superset.db_engine_specs.mysql.MySQLEngineSpec'>,
 'odelasticsearch': <class 'superset.db_engine_specs.elasticsearch.OpenDistroEngineSpec'>,
 'oracle': <class 'superset.db_engine_specs.oracle.OracleEngineSpec'>,
 'pinot': <class 'superset.db_engine_specs.pinot.PinotEngineSpec'>,
 'postgres': <class 'superset.db_engine_specs.postgres.PostgresEngineSpec'>,
 'postgresql': <class 'superset.db_engine_specs.postgres.PostgresEngineSpec'>,
 'presto': <class 'superset.db_engine_specs.presto.PrestoEngineSpec'>,
 'redshift': <class 'superset.db_engine_specs.redshift.RedshiftEngineSpec'>,
 'snowflake': <class 'superset.db_engine_specs.snowflake.SnowflakeEngineSpec'>,
 'solr': <class 'superset.db_engine_specs.solr.SolrEngineSpec'>,
 'sqlite': <class 'superset.db_engine_specs.sqlite.SqliteEngineSpec'>,
 'teradata': <class 'superset.db_engine_specs.teradata.TeradataEngineSpec'>,
 'trino': <class 'superset.db_engine_specs.trino.TrinoEngineSpec'>,
 'vertica': <class 'superset.db_engine_specs.vertica.VerticaEngineSpec'>}
```

The only difference is the absence of the base class:` 'base': <class 'superset.db_engine_specs.base.BaseEngineSpec'>,`

### TEST PLAN

Tested Superset, everything works as expected.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -62,7 +62,43 @@ def get_git_sha():
...
+            "presto=superset.db_engine_specs.presto:PrestoEngineSpec",
+            "redshift=superset.db_engine_specs.redshift:RedshiftEngineSpec",
+            "snowflake=superset.db_engine_specs.snowflake:SnowflakeEngineSpec",
+            "solr=superset.db_engine_specs.solr:SolrEngineSpec",
+            "sqlite=superset.db_engine_specs.sqlite:SqliteEngineSpec",
+            "teradata=superset.db_engine_specs.teradata:TeradataEngineSpec",
+            "trino=superset.db_engine_specs.trino:TrinoEngineSpec",
+            "vertica=superset.db_engine_specs.vertica:VerticaEngineSpec",
+        ],
+    },
```
Comment (villebro) [inline setup.py:101]: Would it be possible to automate this part by iterating over modules in `superset/db_engine_specs/*.py` and checking that they are `issubclass` but not `isinstance` of `BaseEngineSpec`?

**Review by villebro — APPROVED** [review]
This is great! One question about automating population of the built-in specs, other than that LGTM!

File: setup.py (hunk 1 above)
Comment (betodealmeida) [reply setup.py]: Will do!

File: superset/db_engine_specs/__init__.py [hunk 2]
```diff
@@ -28,28 +28,57 @@
...
+        )
 
-for (_, name, _) in pkgutil.iter_modules([Path(__file__).parent]):  # type: ignore
-    imported_module = import_module("." + name, package=__name__)
+    # load additional engines from external modules
+    for ep in iter_entry_points("superset.db_engine_specs"):
+        try:
+            engine_spec = ep.load()
+        except Exception:  # pylint: disable=broad-except
+            logger.warning("Unable to load engine spec: %s", engine_spec)
```
Comment (john-bodley) [inline superset/db_engine_specs/__init__.py:70]: @betodealmeida there's a bug here. If line 68 throws an exception then `engine_spec` isn't defined.

**Pull request merged**